File: src/backend/routes/connections.ts
import { FastifyInstance } from 'fastify';
import { MongoClient } from 'mongodb';
import { ConnectionStorage, StoredConnection } from '../storage/connections.js';
import { PasswordStorage } from '../storage/keyring.js';
import { ConnectionPool } from '../db/mongodb.js';
import {
  CreateConnectionRequest,
  UpdateConnectionRequest,
  ConnectionResponse,
  TestConnectionRequest,
  TestConnectionResponse,
  ConnectRequest,
} from '../../shared/contracts.js';

export interface ConnectionRoutesOptions {
  storage: ConnectionStorage;
  passwords: PasswordStorage;
  pool: ConnectionPool;
}

/** Strip credentials from a MongoDB URI, returning the cleaned URI plus extracted username/password. */
export function stripCredentials(uri: string): {
  strippedUri: string;
  username: string;
  password: string;
} {
  const parsed = new URL(uri);
  const username = decodeURIComponent(parsed.username);
  const password = decodeURIComponent(parsed.password);

  // Remove credentials from the URL
  parsed.username = '';
  parsed.password = '';

  // URL.toString() re-encodes, so we build from parts to avoid double-encoding
  let strippedUri = `${parsed.protocol}//`;
  strippedUri += parsed.host;
  strippedUri += parsed.pathname;
  if (parsed.search) strippedUri += parsed.search;

  return { strippedUri, username, password };
}

/** Re-insert credentials into a credential-stripped URI for connecting. */
export function injectCredentials(uri: string, username: string, password: string): string {
  if (!username) return uri;

  const parsed = new URL(uri);
  parsed.username = encodeURIComponent(username);
  if (password) {
    parsed.password = encodeURIComponent(password);
  }

  // Rebuild to avoid URL class encoding issues
  let result = `${parsed.protocol}//`;
  result += encodeURIComponent(username);
  if (password) result += ':' + encodeURIComponent(password);
  result += '@';
  result += parsed.host;
  result += parsed.pathname;
  if (parsed.search) result += parsed.search;

  return result;
}

/** Extract database name from a MongoDB URI. */
function getDatabaseFromUri(uri: string): string | undefined {
  try {
    const parsed = new URL(uri);
    const db = parsed.pathname.slice(1); // remove leading /
    return db || undefined;
  } catch {
    return undefined;
  }
}

function toConnectionResponse(
  conn: StoredConnection & { error?: string },
  isConnected: boolean
): ConnectionResponse {
  return {
    id: conn.id,
    name: conn.name,
    uri: conn.uri,
    username: conn.username,
    savePassword: conn.savePassword,
    isConnected,
    createdAt: conn.createdAt,
    updatedAt: conn.updatedAt,
    error: conn.error,
  };
}

export async function connectionRoutes(
  fastify: FastifyInstance,
  opts: ConnectionRoutesOptions
): Promise<void> {
  const { storage, passwords, pool } = opts;

  // List all connections
  fastify.get('/', async (_request, reply) => {
    const connections = await storage.list();
    const response: ConnectionResponse[] = connections.map((conn) =>
      toConnectionResponse(conn, pool.isConnected(conn.id))
    );
    return reply.send(response);
  });

  // Get single connection
  fastify.get<{ Params: { id: string } }>('/:id', async (request, reply) => {
    const conn = await storage.get(request.params.id);
    if (!conn) {
      return reply.status(404).send({
        error: 'NotFound',
        message: `Connection '${request.params.id}' not found`,
        statusCode: 404,
      });
    }
    return reply.send(toConnectionResponse(conn, pool.isConnected(conn.id)));
  });

  // Create connection
  fastify.post<{ Body: CreateConnectionRequest }>('/', async (request, reply) => {
    const { name, uri, savePassword = true } = request.body;

    // Strip credentials from the URI before storing
    const { strippedUri, username, password } = stripCredentials(uri);

    const conn = await storage.create({
      name,
      uri: strippedUri,
      username: username || undefined,
      savePassword,
    });

    if (password && savePassword) {
      await passwords.set(conn.id, password);
    }

    return reply.status(201).send(toConnectionResponse(conn, false));
  });

  // Update connection
  fastify.put<{ Params: { id: string }; Body: UpdateConnectionRequest }>(
    '/:id',
    async (request, reply) => {
      const { name, uri, savePassword } = request.body;

      try {
        const updates: Partial<Omit<StoredConnection, 'id' | 'createdAt'>> = {};

        if (name !== undefined) {
          updates.name = name;
        }

        if (savePassword !== undefined) {
          updates.savePassword = savePassword;
        }

        // Determine effective savePassword: use request value, fall back to stored value
        let effectiveSavePassword = savePassword;
        if (effectiveSavePassword === undefined) {
          const existing = await storage.get(request.params.id);
          effectiveSavePassword = existing?.savePassword ?? true;
        }

        if (uri !== undefined) {
          const { strippedUri, username, password } = stripCredentials(uri);
          updates.uri = strippedUri;
          updates.username = username || undefined;

          if (password && effectiveSavePassword) {
            await passwords.set(request.params.id, password);
          } else {
            // No password, or savePassword is false — clear stored password
            await passwords.delete(request.params.id);
          }
        } else if (savePassword === false) {
          // URI not changed but savePassword flipped to false — clear keyring
          await passwords.delete(request.params.id);
        }

        const conn = await storage.update(request.params.id, updates);

        return reply.send(toConnectionResponse(conn, pool.isConnected(conn.id)));
      } catch (e) {
        const error = e as Error;
        if (error.message.includes('not found')) {
          return reply.status(404).send({
            error: 'NotFound',
            message: error.message,
            statusCode: 404,
          });
        }
        throw e;
      }
    }
  );

  // Delete connection
  fastify.delete<{ Params: { id: string } }>('/:id', async (request, reply) => {
    const id = request.params.id;

    // Disconnect if connected
    if (pool.isConnected(id)) {
      await pool.disconnect(id);
    }

    try {
      await storage.delete(id);
      await passwords.delete(id);
      return reply.status(204).send();
    } catch (e) {
      const error = e as Error;
      if (error.message.includes('not found')) {
        return reply.status(404).send({
          error: 'NotFound',
          message: error.message,
          statusCode: 404,
        });
      }
      throw e;
    }
  });

  // Test unsaved connection (must be before /:id/test to match correctly)
  fastify.post<{ Body: TestConnectionRequest }>('/test', async (request, reply) => {
    const { uri } = request.body;

    const startTime = Date.now();
    const client = new MongoClient(uri, {
      serverSelectionTimeoutMS: 5000,
      connectTimeoutMS: 5000,
    });

    try {
      await client.connect();
      await client.db('admin').command({ ping: 1 });
      const latencyMs = Date.now() - startTime;

      const response: TestConnectionResponse = {
        success: true,
        message: 'Connection successful',
        latencyMs,
      };
      return reply.send(response);
    } catch (e) {
      const error = e as Error;
      const response: TestConnectionResponse = {
        success: false,
        message: error.message,
      };
      return reply.send(response);
    } finally {
      await client.close();
    }
  });

  // Test saved connection by ID
  fastify.post<{ Params: { id: string }; Body: { password?: string } }>(
    '/:id/test',
    async (request, reply) => {
      const conn = await storage.get(request.params.id);
      if (!conn) {
        return reply.status(404).send({
          error: 'NotFound',
          message: `Connection '${request.params.id}' not found`,
          statusCode: 404,
        });
      }

      if (conn.error) {
        return reply.status(400).send({
          error: 'BadRequest',
          message: conn.error,
          statusCode: 400,
        });
      }

      const bodyPassword = (request.body as { password?: string } | undefined)?.password;
      const password = bodyPassword ?? (await passwords.get(conn.id));
      const uri = conn.username
        ? injectCredentials(conn.uri, conn.username, password ?? '')
        : conn.uri;

      const startTime = Date.now();
      const client = new MongoClient(uri, {
        serverSelectionTimeoutMS: 5000,
        connectTimeoutMS: 5000,
      });

      try {
        await client.connect();
        await client.db('admin').command({ ping: 1 });
        const latencyMs = Date.now() - startTime;

        const response: TestConnectionResponse = {
          success: true,
          message: 'Connection successful',
          latencyMs,
        };
        return reply.send(response);
      } catch (e) {
        const error = e as Error;
        const response: TestConnectionResponse = {
          success: false,
          message: error.message,
        };
        return reply.send(response);
      } finally {
        await client.close();
      }
    }
  );

  // Connect
  fastify.post<{ Params: { id: string }; Body: ConnectRequest }>(
    '/:id/connect',
    async (request, reply) => {
      const id = request.params.id;
      const body = (request.body ?? {}) as ConnectRequest;

      // Force-disconnect stale entry so reconnect always works
      if (pool.isConnected(id)) {
        await pool.forceDisconnect(id);
      }

      let conn = await storage.get(id);
      if (!conn) {
        return reply.status(404).send({
          error: 'NotFound',
          message: `Connection '${id}' not found`,
          statusCode: 404,
        });
      }

      if (conn.error) {
        return reply.status(400).send({
          error: 'BadRequest',
          message: conn.error,
          statusCode: 400,
        });
      }

      // "Remember password" flow: persist password to keyring and update connection
      if (body.savePassword && body.password) {
        await passwords.set(id, body.password);
        conn = await storage.update(id, { savePassword: true });
      }

      const password = body.password ?? (await passwords.get(conn.id));
      const uri = conn.username
        ? injectCredentials(conn.uri, conn.username, password ?? '')
        : conn.uri;
      const database = getDatabaseFromUri(conn.uri);

      try {
        await pool.connect(id, { uri, database });
        return reply.send(toConnectionResponse(conn, true));
      } catch (e) {
        const error = e as Error;
        return reply.status(500).send({
          error: 'ConnectionFailed',
          message: error.message,
          statusCode: 500,
        });
      }
    }
  );

  // Disconnect
  fastify.post<{ Params: { id: string } }>('/:id/disconnect', async (request, reply) => {
    const id = request.params.id;

    const conn = await storage.get(id);
    if (!conn) {
      return reply.status(404).send({
        error: 'NotFound',
        message: `Connection '${id}' not found`,
        statusCode: 404,
      });
    }

    // Use forceDisconnect to handle both active and stale connections
    await pool.forceDisconnect(id);
    return reply.send(toConnectionResponse(conn, false));
  });
}

File: src-tauri/src/commands/connections.rs
use serde::{Deserialize, Serialize};
use std::time::Instant;
use tauri::State;
use tokio_util::sync::CancellationToken;

use crate::credentials::{get_database_from_uri, inject_credentials, strip_credentials};
use crate::error::DgridError;
use crate::pool::MongoConnectionOptions;
use crate::state::AppState;
use crate::storage::{CreateConnectionInput, StoredConnection, UpdateConnectionInput};

// --- Request/Response types (matching contracts.ts) ---

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateConnectionRequest {
    pub name: String,
    pub uri: String,
    #[serde(default = "default_true")]
    pub save_password: bool,
}

fn default_true() -> bool {
    true
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateConnectionRequest {
    pub name: Option<String>,
    pub uri: Option<String>,
    pub save_password: Option<bool>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ConnectionResponse {
    pub id: String,
    pub name: String,
    pub uri: String,
    pub username: Option<String>,
    pub save_password: bool,
    pub is_connected: bool,
    pub created_at: String,
    pub updated_at: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TestConnectionRequest {
    pub uri: String,
    pub operation_id: Option<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TestConnectionResponse {
    pub success: bool,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub latency_ms: Option<u64>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConnectRequest {
    pub password: Option<String>,
    pub save_password: Option<bool>,
}

// --- Helper ---

fn to_connection_response(conn: &StoredConnection, is_connected: bool) -> ConnectionResponse {
    ConnectionResponse {
        id: conn.id.clone(),
        name: conn.name.clone(),
        uri: conn.uri.clone(),
        username: conn.username.clone(),
        save_password: conn.save_password,
        is_connected,
        created_at: conn.created_at.clone(),
        updated_at: conn.updated_at.clone(),
        error: conn.error.clone(),
    }
}

// --- Commands ---

#[tauri::command]
pub async fn list_connections(
    state: State<'_, AppState>,
) -> Result<Vec<ConnectionResponse>, DgridError> {
    let connections = {
        let storage = state.storage.lock().unwrap();
        storage.list()?
    };

    let mut responses = Vec::new();
    for conn in &connections {
        let is_connected = state.pool.is_connected(&conn.id).await;
        responses.push(to_connection_response(conn, is_connected));
    }
    Ok(responses)
}

#[tauri::command]
pub async fn get_connection(
    state: State<'_, AppState>,
    id: String,
) -> Result<ConnectionResponse, DgridError> {
    let conn = {
        let storage = state.storage.lock().unwrap();
        storage.get(&id)?
    };

    let conn = conn.ok_or_else(|| DgridError::NotFound {
        entity: "Connection".into(),
        id: id.clone(),
    })?;

    let is_connected = state.pool.is_connected(&id).await;
    Ok(to_connection_response(&conn, is_connected))
}

#[tauri::command]
pub async fn create_connection(
    state: State<'_, AppState>,
    request: CreateConnectionRequest,
) -> Result<ConnectionResponse, DgridError> {
    let stripped = strip_credentials(&request.uri);

    let conn = {
        let storage = state.storage.lock().unwrap();
        storage.create(CreateConnectionInput {
            name: request.name,
            uri: stripped.stripped_uri,
            username: if stripped.username.is_empty() {
                None
            } else {
                Some(stripped.username.clone())
            },
            save_password: request.save_password,
        })?
    };

    if !stripped.password.is_empty() && request.save_password {
        state.passwords.set(&conn.id, &stripped.password)?;
    }

    Ok(to_connection_response(&conn, false))
}

#[tauri::command]
pub async fn update_connection(
    state: State<'_, AppState>,
    id: String,
    request: UpdateConnectionRequest,
) -> Result<ConnectionResponse, DgridError> {
    let mut input = UpdateConnectionInput {
        name: request.name,
        uri: None,
        username: None,
        save_password: request.save_password,
    };

    // Determine effective savePassword
    let effective_save_password = match request.save_password {
        Some(sp) => sp,
        None => {
            let storage = state.storage.lock().unwrap();
            storage.get(&id)?.map(|c| c.save_password).unwrap_or(true)
        }
    };

    if let Some(uri) = &request.uri {
        let stripped = strip_credentials(uri);
        input.uri = Some(stripped.stripped_uri);
        input.username = Some(if stripped.username.is_empty() {
            None
        } else {
            Some(stripped.username)
        });

        if !stripped.password.is_empty() && effective_save_password {
            state.passwords.set(&id, &stripped.password)?;
        } else {
            state.passwords.delete(&id);
        }
    } else if request.save_password == Some(false) {
        state.passwords.delete(&id);
    }

    let conn = {
        let storage = state.storage.lock().unwrap();
        storage.update(&id, input)?
    };

    let is_connected = state.pool.is_connected(&id).await;
    Ok(to_connection_response(&conn, is_connected))
}

#[tauri::command]
pub async fn delete_connection(state: State<'_, AppState>, id: String) -> Result<(), DgridError> {
    if state.pool.is_connected(&id).await {
        state.pool.disconnect(&id).await?;
    }

    {
        let storage = state.storage.lock().unwrap();
        storage.delete(&id)?;
    }

    state.passwords.delete(&id);
    Ok(())
}

#[tauri::command]
pub async fn test_connection(
    state: State<'_, AppState>,
    request: TestConnectionRequest,
) -> Result<TestConnectionResponse, DgridError> {
    execute_test_connection(&state, request.uri, request.operation_id).await
}

async fn execute_test_connection(
    state: &AppState,
    uri: String,
    operation_id: Option<String>,
) -> Result<TestConnectionResponse, DgridError> {
    let start = Instant::now();

    let token = CancellationToken::new();
    if let Some(ref operation_id) = operation_id {
        let mut tokens = state.test_connection_tokens.write().await;
        if let Some(old) = tokens.remove(operation_id) {
            old.cancel();
        }
        tokens.insert(operation_id.clone(), token.clone());
    }

    let result = tokio::select! {
        result = async {
            let client_options = mongodb::options::ClientOptions::parse(&uri)
                .await
                .map_err(|e| DgridError::Connection(e.to_string()))?;

            let mut opts = client_options;
            opts.server_selection_timeout = Some(std::time::Duration::from_secs(5));
            opts.connect_timeout = Some(std::time::Duration::from_secs(5));

            let client = mongodb::Client::with_options(opts)
                .map_err(|e| DgridError::Connection(e.to_string()))?;

            match client
                .database("admin")
                .run_command(mongodb::bson::doc! { "ping": 1 })
                .await
            {
                Ok(_) => {
                    let latency_ms = start.elapsed().as_millis() as u64;
                    Ok(TestConnectionResponse {
                        success: true,
                        message: "Connection successful".to_string(),
                        latency_ms: Some(latency_ms),
                    })
                }
                Err(e) => Ok(TestConnectionResponse {
                    success: false,
                    message: e.to_string(),
                    latency_ms: None,
                }),
            }
        } => result,
        _ = token.cancelled() => Err(DgridError::TestConnectionCancelled),
    };

    if let Some(ref operation_id) = operation_id {
        let mut tokens = state.test_connection_tokens.write().await;
        tokens.remove(operation_id);
    }

    result
}

#[tauri::command]
pub async fn cancel_test_connection(
    state: State<'_, AppState>,
    operation_id: String,
) -> Result<(), DgridError> {
    let mut tokens = state.test_connection_tokens.write().await;
    if let Some(token) = tokens.remove(&operation_id) {
        token.cancel();
    }
    Ok(())
}

#[tauri::command]
pub async fn test_saved_connection(
    state: State<'_, AppState>,
    id: String,
    password: Option<String>,
    operation_id: Option<String>,
) -> Result<TestConnectionResponse, DgridError> {
    let conn = {
        let storage = state.storage.lock().unwrap();
        storage.get(&id)?
    };

    let conn = conn.ok_or_else(|| DgridError::NotFound {
        entity: "Connection".into(),
        id: id.clone(),
    })?;

    if let Some(err) = &conn.error {
        return Err(DgridError::Validation(err.clone()));
    }

    let resolved_password = password.or_else(|| state.passwords.get(&conn.id));
    let uri = if let Some(username) = &conn.username {
        inject_credentials(&conn.uri, username, &resolved_password.unwrap_or_default())
    } else {
        conn.uri.clone()
    };

    execute_test_connection(&state, uri, operation_id).await
}

#[tauri::command]
pub async fn connect_to_connection(
    state: State<'_, AppState>,
    id: String,
    request: Option<ConnectRequest>,
) -> Result<ConnectionResponse, DgridError> {
    let request = request.unwrap_or(ConnectRequest {
        password: None,
        save_password: None,
    });

    // Force-disconnect stale entry so reconnect always works
    if state.pool.is_connected(&id).await {
        state.pool.force_disconnect(&id).await;
    }

    let conn = {
        let storage = state.storage.lock().unwrap();
        storage.get(&id)?
    };

    let mut conn = conn.ok_or_else(|| DgridError::NotFound {
        entity: "Connection".into(),
        id: id.clone(),
    })?;

    if let Some(err) = &conn.error {
        return Err(DgridError::Validation(err.clone()));
    }

    // "Remember password" flow
    if request.save_password == Some(true) {
        if let Some(ref password) = request.password {
            state.passwords.set(&id, password)?;
            let storage = state.storage.lock().unwrap();
            conn = storage.update(
                &id,
                UpdateConnectionInput {
                    name: None,
                    uri: None,
                    username: None,
                    save_password: Some(true),
                },
            )?;
        }
    }

    let resolved_password = request.password.or_else(|| state.passwords.get(&conn.id));
    let uri = if let Some(username) = &conn.username {
        inject_credentials(&conn.uri, username, &resolved_password.unwrap_or_default())
    } else {
        conn.uri.clone()
    };
    let database = get_database_from_uri(&conn.uri);

    let token = CancellationToken::new();
    {
        let mut tokens = state.connection_tokens.write().await;
        if let Some(old) = tokens.remove(&id) {
            old.cancel();
        }
        tokens.insert(id.clone(), token.clone());
    }

    let result = tokio::select! {
        result = state.pool.connect(&id, MongoConnectionOptions { uri, database }) => result,
        _ = token.cancelled() => Err(DgridError::ConnectionCancelled),
    };

    {
        let mut tokens = state.connection_tokens.write().await;
        tokens.remove(&id);
    }

    if matches!(result, Err(DgridError::ConnectionCancelled)) {
        state.pool.force_disconnect(&id).await;
    }

    result?;

    Ok(to_connection_response(&conn, true))
}

#[tauri::command]
pub async fn cancel_connect_to_connection(
    state: State<'_, AppState>,
    id: String,
) -> Result<(), DgridError> {
    let mut tokens = state.connection_tokens.write().await;
    if let Some(token) = tokens.remove(&id) {
        token.cancel();
    }
    Ok(())
}

#[tauri::command]
pub async fn disconnect_from_connection(
    state: State<'_, AppState>,
    id: String,
) -> Result<ConnectionResponse, DgridError> {
    let conn = {
        let storage = state.storage.lock().unwrap();
        storage.get(&id)?
    };

    let conn = conn.ok_or_else(|| DgridError::NotFound {
        entity: "Connection".into(),
        id: id.clone(),
    })?;

    state.pool.force_disconnect(&id).await;
    Ok(to_connection_response(&conn, false))
}
